### ISSUE TITLE
colmapWithCuda: build failure

### ISSUE THREAD

**Issue by SomeoneSerge** [NixOS/nixpkgs#275113]
### Describe the bug

Build fails at link time: https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fsz5rrd5k90incm69risxqx7pkcnb2xih-colmap-3.7.drv/log?via-job=9e2e3781-42da-4528-878e-8c36076d1d55

### Steps To Reproduce

`nix build -f ./. --arg config '{ cudaSupport = true; allowUnfree = true;  }' -L colmap`

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by SomeoneSerge** [comment]
> but I get C++ ABI unknown reference errors on a few packages.

I suspect colmap needs this: https://github.com/NixOS/nixpkgs/blob/40e221feff8a3cd68c683ef6e5f0bed3c769bace/pkgs/development/libraries/opencv/4.x.nix#L90-L93

**Comment by kowale** [comment]
Thank you, that's definitely more elegant. I ran it for GCC11 (which is still required by Colmap), it didn't rebuild as much and I didn't need the test patches.

```nix
{ pkgs, ... }:

let

    # https://github.com/NixOS/nixpkgs/issues/275113
    colmap = pkgs.colmap.override { mkDerivation = pkgs.cudaPackages_11.backendStdenv.mkDerivation; };
in

    colmap.overrideAttrs (final: prev: rec {

        version = "3.9.1";

        src = pkgs.fetchFromGitHub {
            owner = "colmap";
            repo = "colmap";
            rev = version;
            hash = "sha256-Xb4JOttCMERwPYs5DyGKHw+f9Wik1/rdJQKbgVuygH8=";
        };

        buildInputs = with pkgs; prev.buildInputs ++ [
            flann
            cgal
            gmp
            mpfr
            xorg.libSM
        ];

        nativeBuildInputs = with pkgs; prev.nativeBuildInputs ++ [
            qt5.wrapQtAppsHook
        ];

        # https://github.com/colmap/colmap/issues/1940
        cmakeFlags =
            [
                "-DCUDA_ENABLED=ON"
                "-DCMAKE_CUDA_ARCHITECTURES=86"
            ];
    })
```

I will read into contributing policy and try to make a PR sometime soon. Do you have any suggestions on cudaCapabilities? I couldn't find this config used anywhere (just the default in `cuda-modules/flags.nix`), but it seems like it should not be hardcoded right?

**Comment by SomeoneSerge** [comment]
> I will read into contributing policy and try to make a PR sometime soon. Do you have any suggestions on cudaCapabilities? I couldn't find this config used anywhere (just the default in cuda-modules/flags.nix), but it seems like it should not be hardcoded right?

Correct. Grep nixpkgs for `CMAKE_CUDA_ARCHITECTURES`, e.g. look at the nix expressions for `cudaPackages.saxpy`, `python3Packages.torch`, `opencv4`, etc

**Comment by SomeoneSerge** [comment]
> . I ran it for GCC11 (which is still required by Colmap)

Do you mean `cudaPackages_11`? That uses gcc11 btw.
